### PR TITLE
Added functionality for registration of AnalysisNotebookResults.

### PR DESCRIPTION
**Pull request by MWolfR**

### CODE REVIEW

File: src/obi_notebook/register_notebook_result.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+            entity_id=nb_entity_register.id,
+            entity_type=models.AnalysisNotebookResult,
+            file_path=notebook_file,
+            file_name=path.split(notebook_file)[1],
+            file_content_type=types.ContentType.application_x_ipynb_json,
+            asset_label=types.AssetLabel.jupyter_notebook,
+        )
+        if asset_obj.id is not None:
+            print("...NotebookResult successfully created!")
+            return nb_entity_register.id, asset_obj.id
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: If `client.upload_file` fails it will not return and asset with id=None, but rather raise.

File: src/obi_notebook/register_notebook_result.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+from IPython.display import clear_output, display
+from obi_auth import get_user_info
+
+_IPYNB = ".ipynb"
+
+
+def _find_potential_notebook_files() -> list[str]:
+    """Find .ipynb files in the the current working directory."""
+    fns = [fn for fn in listdir() if path.splitext(fn)[1] == _IPYNB]
+    return fns
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: This can be simplified by importing glob and:
```suggestion
    return list(glob.glob("*.ipynb"))
```

File: src/obi_notebook/register_notebook_result.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+    pretty_str = "Last saved: "
+    if days_delta > 0:
+        pretty_str = pretty_str + f"{days_delta} day(s) "
+    if hours_delta > 0:
+        pretty_str = pretty_str + f"{hours_delta} hour(s) "
+    if minutes_delta > 0:
+        pretty_str = pretty_str + f"{minutes_delta} minute(s) "
+    pretty_str = pretty_str + f"{seconds_delta} second(s) ago"
+    return pretty_str
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: Minor: Slightly more readable:
```suggestion
    modified = datetime.fromtimestamp(Path(fn).stat().st_mtime)
    delta = datetime.now() - modified

    seconds = int(delta.total_seconds())

    days, seconds = divmod(seconds, 86400)
    hours, seconds = divmod(seconds, 3600)
    minutes, seconds = divmod(seconds, 60)

    parts = []
    if days:
        parts.append(f"{days} day(s)")
    if hours:
        parts.append(f"{hours} hour(s)")
    if minutes:
        parts.append(f"{minutes} minute(s)")

    parts.append(f"{seconds} second(s) ago")

    return "Last saved: " + " ".join(parts)
```

File: src/obi_notebook/register_notebook_result.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+
+def _this_user(client: Client) -> str | None:
+    """Get the authenticated user's "sub_id" for use with queries."""
+    user_info = get_user_info(client._token_manager.get_token())
+    return user_info.get("sub", None)
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: Is it possible to get users without sub? @danifr

File: src/obi_notebook/register_notebook_result.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+        query = {"name": notebook_name, "authorized_public": False}
+    else:
+        query = {"created_by__sub_id": user_sub, "name": notebook_name, "authorized_public": False}
+    query_res = client.search_entity(entity_type=models.AnalysisNotebookResult, query=query)
+    return query_res.all()
+
+
+def _entity_str(entity: models.AnalysisNotebookResult) -> str:
+    """For a NotebookResult entity, create a descriptive string."""
+    if entity.update_date is not None:
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: update_date cannot be None. At registration it is given the same value as creation_date.

File: src/obi_notebook/register_notebook_result.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        ):
+            print("Updating name and description...")
+            client.update_entity(
+                entity_id=nb_entity_found.id,
+                entity_type=models.AnalysisNotebookResult,
+                attrs_or_entity={"name": notebook_name, "description": notebook_description},
+            )
+        assets = [
+            _x for _x in nb_entity_found.assets if _x.label == types.AssetLabel.jupyter_notebook
+        ]
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]:
```suggestion
        assets = client.select_assets(
            entity=nb_entity_found,
            selection={"label": types.AssetLabel.jupyter_notebook}
        ).all()
```

File: src/obi_notebook/register_notebook_result.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    entity_id: UUID,
+    notebook_file: Path,
+    notebook_name: str,
+    notebook_description: str,
+):
+    """Update an existing NotebookResult entity (name, description and assets)."""
+    nb_entity_found = client.get_entity(
+        entity_id=entity_id, entity_type=models.AnalysisNotebookResult
+    )
+    if nb_entity_found.id is not None:
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: An entity returned by get_entity has always an id.

File: src/obi_notebook/register_notebook_result.py (hunk 4 above)
Comment (danifr) [reply src/obi_notebook/register_notebook_result.py]: Short answer: No, it is not possible

Long answer: AFAIK the `sub` claim is included by default in all access tokens minted by Keycloak. Removing the `sub` claim would require some custom token generation logic or customization of Keycloak, which I don't think we will ever implement :)

File: src/obi_notebook/register_notebook_result.py (hunk 1 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: I know. The reason I test for None is simply that I want to guarantee in this function that I return a tuple of UUIDs, and entitysdk only specifies that it returns an `Asset` where _technically_ the .id can be None.  

If the upload fails, an `EntitySDKError` is raised and not caught and that is intentional.

File: src/obi_notebook/register_notebook_result.py (hunk 5 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: Again. I added it due to the code inspection in Visual Studio. All it knows is that the `.update_date` of an entity is declared to be `datetime | None`. So, if I do not add that test, it marks it red because `None` has no function `strftime`.

If you think it should still be removed, I can remove the test, but I see no harm in having it.

File: src/obi_notebook/register_notebook_result.py (hunk 6 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: Oh, I did not know of that function. Good to know; thanks.

Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: Applied in update.

**Comment by MWolfR** [comment]
If you insist, I will remove the superfluous checks against `None`. 
I just added them to make Visual Studio happy in the first place. Let me know...

File: src/obi_notebook/register_notebook_result.py (hunk 5 above)
Comment (eleftherioszisis) [reply src/obi_notebook/register_notebook_result.py]: Does the following fix the linting issue?
```
from typing import cast
update_date = cast(datetime, entity.update_date)
```
Unfortunately, there is no good way to use an entity both for read and registration/update, keep the same type, and have typing know that the schema is conditionally different.

IMO it is better to use a cast because it shows the intent that we know this always exists and it is a typing issue to ignore.

File: src/obi_notebook/register_notebook_result.py (hunk 1 above)
Comment (eleftherioszisis) [reply src/obi_notebook/register_notebook_result.py]: Yeah, same issue. When the asset is used in registration (e.g. uploading a file) it is not possible to be returned without an id because the server will always return one in the schema. I would also cast the offending types with `cast(ID, X)` where ID is imported from entitysdk.types.

File: src/obi_notebook/register_notebook_result.py (hunk 4 above)
Comment (eleftherioszisis) [reply src/obi_notebook/register_notebook_result.py]: In that case we can safely return `user_info["sub"]`.

File: src/obi_notebook/register_notebook_result.py (hunk 1 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: I confirm it works in the update using `cast`.

File: src/obi_notebook/register_notebook_result.py (hunk 4 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: Changed to `user_info["sub"]` accordingly.

File: src/obi_notebook/register_notebook_result.py (hunk 5 above)
Comment (MWolfR) [reply src/obi_notebook/register_notebook_result.py]: Done and works.

**Comment by MWolfR** [comment]
I removed the superfluous `None` checks as discussed. Now all objections should be fixed. Can you check again, @eleftherioszisis ?

File: src/obi_notebook/register_notebook_result.py [hunk 8]
```diff
@@ -109,36 +110,36 @@ def _update_result(
...
+            entity_id=entity_id,
+            entity_type=models.AnalysisNotebookResult,
+            asset_id=asset.id,
+        )
+    print("Uploding new asset...")
+    asset_obj = client.upload_file(
+        entity_id=entity_id,
+        entity_type=models.AnalysisNotebookResult,
+        file_path=notebook_file,
+        file_name=path.split(notebook_file)[1],
```
Comment (eleftherioszisis) [inline src/obi_notebook/register_notebook_result.py]: If notebook_file is a Path as the type implies, you can do `notebook_file.name` which would extract from `/a/b/c.ext` the `c.ext` for example.

**Review by eleftherioszisis — APPROVED** [review]
